app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt

```diff
@@ -244,19 +248,80 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
 	 * not authorized then the the [QueueMetadata] is created from the app remote API.
 	 */
 	fun createLikedSongsQueueMetadata() {
+		if (createQueueMetadataJob?.isActive == true) {
+			createQueueMetadataJob?.cancel()
+		}
+
 		createQueueMetadataJob = GlobalScope.launch(defaultDispatcher) {
 			queueItems = webApi.getLikedSongs(this@SpotifyAppController) ?: emptyList()
+
 			if (queueItems.isNotEmpty()) {
-				queueMetadata = QueueMetadata("Liked Songs", null, queueItems)
-				loadQueueCoverart()
+				queueMetadata = QueueMetadata(L.MUSIC_LIKED_SONGS_PLAYLIST_NAME, null, queueItems)
+
+				val hashCode = queueItems.hashCode().toString()
+				val likedSongsStateJson = appSettings[AppSettings.KEYS.SPOTIFY_LIKED_SONGS_PLAYLIST_STATE]
+				val likedSongsState: LikedSongsState
+				if (likedSongsStateJson.isBlank()) {
+					Log.d(TAG, "No previous liked songs state found.")
+					val uri = webApi.createPlaylist(SpotifyWebApi.LIKED_SONGS_PLAYLIST_NAME)
+					if (uri == null) {
+						Log.e(TAG, "Error creating liked songs playlist, falling back to app remote API")
+						queueItems = emptyList()
+						createQueueMetadata(PlayerContext(queueUri, L.MUSIC_LIKED_SONGS_PLAYLIST_NAME, null, null))
+						return@launch
+					}
+					webApi.addSongsToPlaylist(uri.id, queueItems)
+					likedSongsState = LikedSongsState(hashCode, uri.uri, uri.id, null)
+					saveLikedSongsState(likedSongsState)
+				} else {
+					Log.d(TAG, "Found previous liked songs state.")
+					likedSongsState = Json.decodeFromString(likedSongsStateJson)
+
+					if (hashCode != likedSongsState.hashCode) {
+						Log.d(TAG, "Previous liked songs state is no longer valid, updating liked songs playlist and writing new hash.")
+						webApi.replacePlaylistSongs(likedSongsState.playlistId, queueItems)
+						likedSongsState.hashCode = hashCode
+						saveLikedSongsState(likedSongsState)
+					}
+				}
+
+				queueUri = likedSongsState.playlistUri
+
+				if (likedSongsState.queueCoverArtUri == null) {
```

Comment:
It appears that this block is duplicating `loadQueueCoverart` just to persist the `LikedSongsState.queueCoverArtUri`.
I'm not sure that the LikedSongsState's queueCoverArtUri is strictly necessary: The QueueMetadata.coverArt will start out null either way, so the regular loadQueueCoverArt can load the recently-played Liked Songs coverart and update the QueueMetadata like normal? Caching the previous queueCoverArtUri might be incorrect anyways, I suspect the uri changes to show a different picture when the playlist contents change.

Reply:
The scenario around this is a bit interesting as yes the queueCoverArtUri isn't necessary but only in the ideal flow where the user selects the actual Liked Songs playlist. This logic here helps in two ways: 
1. The caching of the cover art uri for the Liked Songs playlist is fast as we don't need to make the extra call to the recently played content API from the Remote API, thus lowering the overall time before the playlist is fully loaded. This is especially useful here as the image uri is always the same static image 
2. If the user selects the dummy Liked Songs playlist, they will then get the cover art that comes from that which is just the 4 cover art images from the first 4 tracks in the playlist by default. We don't want that as we want a single unified experience whether it be the dummy Liked Songs playlist or the real Liked Songs playlist as they should be synonymous and the user shouldn't see a disconnect in that as it might create confusion.

Also the uri for the Liked Songs playlist (the purple to blue gradient background with the white heart image) is always the same for Liked Songs as it is a user collection and I don't believe it is even possible to change it to something else so once that is loaded the initial time that uri will always be correct regardless of the playlist content.

Reply:
Ah good point! Thanks for the explanation.

Reply:
Thinking on this some more: While it would be helpful for filter out the temporary playlist from the browse list so that the user only selects the normal playlist, this flow is still needed to handle the situation where the context has switched to the temporary playlist and then the car app disconnects and reconnects later, still in this temporary context. And so, I view likedSongsState instead as a persistent tempPlaylistState, remembering the original context's title and coverart and the temp playlist identifier for skipping and matching purposes.

I recommend splitting `createLikedSongsQueueMetadata` in two functions: Reinflating the likedSongsState back into a QueueMetadata object (basically these two else blocks here, which will also be needed for temporary Artist Playlists), and the rest of this function, which clones the Liked Songs (well, any current or given playlist really) to the temporary playlist and persists the temp playlist url, name, and coverart for later reinflation. This will remove the need to duplicate the cover art logic, because it'll just persist the currently-loaded cover art (either Liked or Artist) instead of fetching it again here.

Reply:
Ah that's true, I updated the LikedSongsState to be a more generic TemporaryPlaylistState which will allow for greater flexibility in the future when expanding to support Artist playlists. I also agree that the createLikedSongsQueueMetadata() method needed to be split up a bit so I refactored it into the two methods liked you described: one for the creation of a TemporaryPlaylistState and one for the deserialization/inflation of a TemporaryPlaylistState from its JSON representation. I also updated the loadQueueCoverart(..) method so now we can pass in a TemporaryPlaylistState to be updated (and its AppSettings.KEYS key for the save operation) as optional parameters which eliminates the near duplicate code for the fetching of the queue coverart and we can now just keep that logic in one function